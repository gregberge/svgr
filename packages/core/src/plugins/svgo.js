import SVGO from 'svgo'
import cosmiconfig from 'cosmiconfig'
import mergeDeep from 'merge-deep'

const explorer = cosmiconfig('svgo', {
  searchPlaces: [
    'package.json',
    `.svgorc`,
    `.svgorc.json`,
    `.svgorc.yaml`,
    `.svgorc.yml`,
    `svgo.config.js`,
    '.svgo.yml',
  ],
  transform: result => result && result.config,
})

function getBaseSvgoConfig(config) {
  const baseSvgoConfig = { plugins: [] }
  if (config.icon) baseSvgoConfig.plugins.push({ removeViewBox: false })
  return baseSvgoConfig
}

export default async (code, config = {}, state = {}) => {
  if (!config.svgo) return code
  const filePath = state.filePath || process.cwd()
  const svgoRcConfig = await explorer.search(filePath)
  const svgo = new SVGO(
    mergeDeep(getBaseSvgoConfig(config), svgoRcConfig, config.svgoConfig),
  )
  const info = state.filePath
    ? { input: 'file', path: state.filePath }
    : { input: 'string' }
  const { data } = await svgo.optimize(code, info)
  return data
}
