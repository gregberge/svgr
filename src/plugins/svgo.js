import SVGO from 'svgo'
import cosmiconfig from 'cosmiconfig'
import merge from 'lodash/merge'

const explorer = cosmiconfig('svgo', {
  searchPlaces: [
    'package.json',
    `.svgorc`,
    `.svgorc.json`,
    `.svgorc.yaml`,
    `.svgorc.yml`,
    `svgo.config.js`,
    '.svgo.yml',
  ],
  transform: result => result && result.config,
})

export default async (code, opts = {}, state = {}) => {
  const filePath = state.filePath || process.cwd()
  const config = await explorer.search(filePath)
  const svgo = new SVGO(merge({}, config, opts))
  const { data } = await svgo.optimize(code)
  return data
}
