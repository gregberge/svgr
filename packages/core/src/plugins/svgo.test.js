import path from 'path'
import svgo from './svgo'

const baseSvg = `<?xml version="1.0" encoding="UTF-8"?>
<svg width="88px" height="88px" viewBox="0 0 88 88" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
  <!-- Generator: Sketch 46.2 (44496) - http://www.bohemiancoding.com/sketch -->
  <title>Dismiss</title>
  <desc>Created with Sketch.</desc>
  <defs></defs>
  <g id="Blocks" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="square">
      <g id="Dismiss" stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
          <path d="M51,51 L37,37" id="Shape"></path>
      </g>
  </g>
</svg>`

describe('svgo', () => {
  it('should optimize svg', async () => {
    const result = await svgo(baseSvg, { svgo: true })

    expect(result).toMatchSnapshot()
  })

  it('should support config.svgoConfig', async () => {
    const result = await svgo(baseSvg, {
      svgo: true,
      svgoConfig: { plugins: [{ removeDesc: false }] },
    })

    expect(result).toMatchSnapshot()
  })

  it('should support icon with config.svgoConfig plugins', async () => {
    const result = await svgo(baseSvg, {
      svgo: true,
      icon: true,
      svgoConfig: { plugins: [{ removeDesc: false }] },
    })

    expect(result).toMatchSnapshot()
  })

  it('should use state.filePath to detect configuration', async () => {
    const result = await svgo(
      baseSvg,
      { svgo: true },
      { filePath: path.join(__dirname, '../__fixtures__/svgo') },
    )

    expect(result).toMatchSnapshot()
  })

  it('should not remove viewBox with icon option', async () => {
    const result = await svgo(
      baseSvg,
      { icon: true },
      { filePath: path.join(__dirname, '../__fixtures__/svgo') },
    )

    expect(result).toMatchSnapshot()
  })
})
