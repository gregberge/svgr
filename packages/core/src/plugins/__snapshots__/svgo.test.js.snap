// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svgo should not remove viewBox with icon option 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<svg width=\\"88px\\" height=\\"88px\\" viewBox=\\"0 0 88 88\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\">
  <!-- Generator: Sketch 46.2 (44496) - http://www.bohemiancoding.com/sketch -->
  <title>Dismiss</title>
  <desc>Created with Sketch.</desc>
  <defs></defs>
  <g id=\\"Blocks\\" stroke=\\"none\\" stroke-width=\\"1\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"square\\">
      <g id=\\"Dismiss\\" stroke=\\"#063855\\" stroke-width=\\"2\\">
          <path d=\\"M51,37 L37,51\\" id=\\"Shape\\"></path>
          <path d=\\"M51,51 L37,37\\" id=\\"Shape\\"></path>
      </g>
  </g>
</svg>"
`;

exports[`svgo should optimize svg 1`] = `"<svg width=\\"88\\" height=\\"88\\" xmlns=\\"http://www.w3.org/2000/svg\\"><g stroke=\\"#063855\\" stroke-width=\\"2\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"square\\"><path d=\\"M51 37L37 51M51 51L37 37\\"/></g></svg>"`;

exports[`svgo should support config.svgoConfig 1`] = `"<svg width=\\"88\\" height=\\"88\\" xmlns=\\"http://www.w3.org/2000/svg\\"><desc>Created with Sketch.</desc><g stroke=\\"#063855\\" stroke-width=\\"2\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"square\\"><path d=\\"M51 37L37 51M51 51L37 37\\"/></g></svg>"`;

exports[`svgo should use state.filePath to detect configuration 1`] = `"<svg width=\\"88\\" height=\\"88\\" xmlns=\\"http://www.w3.org/2000/svg\\"><desc>Created with Sketch.</desc><g stroke=\\"#063855\\" stroke-width=\\"2\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"square\\"><path d=\\"M51 37L37 51M51 51L37 37\\"/></g></svg>"`;
